makeflow/src/makeflow_cache.c

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright (C) 2014- The University of Notre Dame
+This software is distributed under the GNU General Public License.
+See the file COPYING for details.
+*/
+
+#include "makeflow_cache.h"
+#include "dag.h"
+#include "dag_file.h"
+#include "dag_node.h"
+#include "sha1.h"
+#include "list.h"
+#include "set.h"
+#include "xxmalloc.h"
+#include "stringtools.h"
+#include "batch_job.h"
+#include "debug.h"
+#include "makeflow_log.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+
+void makeflow_cache_generate_id(struct dag_node *n, char *command, struct list*inputs) {
+  struct dag_file *f;
+  char *cache_id = NULL;
+  unsigned char digest[SHA1_DIGEST_LENGTH];
+
+  // add checksum of the node's input files together
+  list_first_item(inputs);
+  while((f = list_next_item(inputs))) {
+    sha1_file(f->filename, digest);
+    cache_id = string_combine(cache_id, sha1_string(digest));
+  }
+
+  sha1_buffer(command, strlen(command), digest);
+  cache_id = string_combine(cache_id, sha1_string(digest));
+  sha1_buffer(cache_id, strlen(cache_id), digest);
+  n -> cache_id = xxstrdup(sha1_string(digest));
+
+  free(cache_id);
+}
+
+void makeflow_cache_populate(struct dag *d, struct dag_node *n, struct list *outputs, struct batch_queue *queue) {
+  char *caching_file_path, *output_file_path, *source_makeflow_file_path, *ancestor_file_path;
+  char *ancestor_cache_id_string = NULL;
+  struct dag_node *ancestor;
+  struct dag_file *f;
+  int sucess;
+  FILE *fp;
+
+  caching_file_path = xxstrdup(d->caching_directory);
+  caching_file_path = string_combine_multi(caching_file_path, n->cache_id, "/outputs", 0);
+  sucess = batch_fs_mkdir(queue,caching_file_path, 0777, 1);
```

Comment:
Actually, for the purposes of caching data, please use fopen, mkdir, on the cache itself, rather than batch_fs_fopen, batch_fs_mkdir, etc.  The latter are only necessary for the files that are in the dag itself.